### PR TITLE
MM-21672: KVCompareAndSet improvements

### PR DESCRIPTION
**Pull request by lieut-data**
#### Summary
As part of developing https://pkg.go.dev/github.com/lieut-data/mattermost-plugin-api@v0.0.6/cluster?tab=doc, I ran into a number of issues with the KV store. One was previously filed as https://mattermost.atlassian.net/browse/MM-21672 noting that `KVCompareAndSet` didn't correctly handle expired keys, but there were other issues:
* `CompareAndDelete` didn't honour expiry
* `List` didn't honour expiry
* Test coverage was not exhaustive

This pull request encompasses the original work captured by https://mattermost.atlassian.net/browse/MM-21672 along with reworking and expanding the unit tests. I strongly recommend reviewing by commit vs. the full changes to help illustrate what's going on.

Unfortunately, these changes won't be live until v5.22 at the earliest, which leaves potential holes in the implementation of the cluster Mutex linked above. I'm separately investigating ways to harden that code to allow it's use with v5.20 and later (the first release which supported `KVCompareAndSet` with expiry via `KVSetWithOptions`). It may even be possible to adopt an "internal" expiration strategy vs. relying on the server at all, further relaxing the dependencies here. But that's a separate investigation.

#### Ticket Link
https://mattermost.atlassian.net/browse/MM-21672

### CODE REVIEW

File: store/sqlstore/plugin_store.go [hunk 1]
```diff
@@ -100,15 +110,18 @@ func (ps SqlPluginStore) CompareAndSet(kv *model.PluginKeyValue, oldValue []byte
 			}
 		}
 	} else {
+		currentTime := model.GetMillis()
+
 		// Update if oldValue is not nil
 		updateResult, err := ps.GetMaster().Exec(
-			`UPDATE PluginKeyValueStore SET PValue = :New, ExpireAt = :ExpireAt WHERE PluginId = :PluginId AND PKey = :Key AND PValue = :Old`,
+			`UPDATE PluginKeyValueStore SET PValue = :New, ExpireAt = :ExpireAt WHERE PluginId = :PluginId AND PKey = :Key AND PValue = :Old AND (ExpireAt = 0 OR ExpireAt > :CurrentTime)`,
```
Comment (cpoile) [inline store/sqlstore/plugin_store.go]: This will leave the expired keys in the store though, right?

File: store/sqlstore/plugin_store.go [hunk 2]
```diff
@@ -60,12 +60,7 @@ func (ps SqlPluginStore) SaveOrUpdate(kv *model.PluginKeyValue) (*model.PluginKe
 		} else if rowsAffected == 0 {
 			// No rows were affected by the update, so let's try an insert
 			if err := ps.GetMaster().Insert(kv); err != nil {
-				// If the error is from unique constraints violation, it's the result of a
-				// valid race and we can report success. Otherwise we have a real error and
-				// need to return it
-				if !IsUniqueConstraintError(err, []string{"PRIMARY", "PluginId", "Key", "PKey", "pkey"}) {
-					return nil, model.NewAppError("SqlPluginStore.SaveOrUpdate", "store.sql_plugin_store.save.app_error", nil, err.Error(), http.StatusInternalServerError)
-				}
+				return nil, model.NewAppError("SqlPluginStore.SaveOrUpdate", "store.sql_plugin_store.save.app_error", nil, err.Error(), http.StatusInternalServerError)
```
Comment (cpoile) [inline store/sqlstore/plugin_store.go]: I'm a bit confused about the reasoning given in the commit message... If the update didn't work, there wasn't a key. If we then try the insert and there was a constraint violation, the row must have been inserted after the update. It seems like returning false is the right thing to do. We "did not successfully save or update the value", but it wasn't an internal server error, right?

**Comment by cpoile** [comment]
As an aside, is `mattermost-plugin-api` ready to use? I can't wait to split workflow off from the server's versions.

File: store/sqlstore/plugin_store.go (hunk 1 above)
Comment (lieut-data) [reply store/sqlstore/plugin_store.go]: That's correct -- we do have a daily background job that prunes expired keys, but from a semantic perspective, the expired key should be treated as if it never existed in the first place, failing the `KVCompareAndSet` when `oldValue != nil`.

File: store/sqlstore/plugin_store.go (hunk 2 above)
Comment (lieut-data) [reply store/sqlstore/plugin_store.go]: You're right that `http.StatusInternalServerError` isn't really the right error code to use here. Perhaps `http.StatusBadRequest`, fully recognizing that we're abusing the status codes altogether here. Thoughts?

Note too we can't actually return `false` from `SaveOrUpdate` like we can (and still do) in `CompareAndSet` under the same circumstances. So I think an error is the right way to model the failure in this case, vs. swallowing and pretending like it was successful.

**Comment by lieut-data** [comment]
> As an aside, is mattermost-plugin-api ready to use? I can't wait to split workflow off from the server's versions.

Sadly, not yet -- at least not outside the new `cluster` package. It hasn't changed much since the hackathon efforts, and we're still trying to evaluate priorities given the team shift going on. I'd love to be able to keep working on it together!!

**Comment by lieut-data** [comment]
@prapti, I'm inclined to rely on the unit test coverage here, with perhaps a smoke test of plugins to ensure continued correct behaviour.

File: store/sqlstore/plugin_store.go (hunk 2 above)
Comment (cpoile) [reply store/sqlstore/plugin_store.go]: In that case, can we use `StatusBadRequest` -- that makes more sense to me. :)

Comment (lieut-data) [reply store/sqlstore/plugin_store.go]: Fixed. I note that I don't explicitly have coverage for this section of the code. At one point in my refactoring, I did -- launched a bunch of inserters and deleters in threads and raced them till I was happy with the outcome, but ran into deadlock/serialization errors from MySQL, which in turn necessitated retry logic, but tuning the right number of retries got tricky... so I ripped that out in favour of forcing the client to deal with it.

Comment (cpoile) [reply store/sqlstore/plugin_store.go]: True, that _would_ be a pain to test. We tested the workflow contention/retry logic by having an action accidentally hold onto the lock for a very long time... :)

**Comment by amyblais** [comment]
@lieut-data Is this and https://github.com/mattermost/mattermost-server/pull/13753 cherry-picked?

**Comment by lieut-data** [comment]
Yes, I just forgot to update the labels in this case. Thanks, @amyblais!

**Pull request merged**